### ISSUE TITLE
How to get list first 5000 files in a list?

### ISSUE THREAD

**Issue by mhghodsi** [vgrem/Office365-REST-Python-Client#226]
I have a list that one of its folders contain more than 15000 files. How would I be able to get the list of all files in that folder?

I am getting this error, 'The attempted operation is prohibited because it exceeds the list view threshold enforced by the administrator.'.

Here is how I am currently getting the list of all files:
 
`libraryRoot = ctx.web.get_folder_by_server_relative_url(list_name)
ctx.load(libraryRoot)
ctx.execute_query()

files = libraryRoot.files
ctx.load(files)
ctx.execute_query()

for index, file in enumerate(files):
    print(f"i : {index} - {file.properties['ServerRelativeUrl']} : Name = {file.properties['Name']} ")`

I'd highly appreciate your help.

**Comment by mhghodsi** [comment]
@vgrem Thanks for the label, I also think this is a bug since the package is unable to handle the request and throws an error.

**Comment by mhghodsi** [comment]
Thanks, @vgrem.

@domdinicola I was wondering if you have any suggestion, highly appreciate your input. Thanks.

**Comment by domdinicola** [comment]
@mhghodsi I think this error is coming from SharePoint, not from this library..
https://www.assistmyteam.com/kb/fix-to-the-attempted-operation-is-prohibited-because-it-exceeds-the-list-view-threshold-enforced-by-the-administrator/

**Comment by mhghodsi** [comment]
@domdinicola Yes, it comes from SharePoint. I was just able to implement the solution and it is rather simple. You need to include "&$skiptoken=Paged=TRUE&$top=5000" to the URL when making the initial request to the API, if the list is larger than 5k rows the response gives you the next URL. If I get a chance I will incorporate it into the package but wanted to share it since you've already contributed into this project. You may close this as I found the solution.

**Comment by vgrem** [comment]
@mhghodsi, indeed, this is the right approach to retrieve data via paged query. Another option would be to optimize query (if filtering or sorting is involved) by introducing column indexes, refer [Manage large lists and libraries in SharePoint](https://support.microsoft.com/en-us/office/manage-large-lists-and-libraries-in-sharepoint-b8588dae-9387-48c2-9248-c24122f07c59) which explains how to create indexes and explain why Listview Threshold error occurs in the first place.

In terms of paged data queries, the good news, the library supports it, refer for instance, [this example](https://github.com/vgrem/Office365-REST-Python-Client/blob/master/examples/sharepoint/lists_and_items/read_large_list.py) on how to retrieve all the list items.

Now going to back to your example, could you please give it another shot by modifying the query from 

```
files = libraryRoot.files
ctx.load(files)
ctx.execute_query()
```

to

```
items = library.items.expand(["Files","Folders"])
ctx.load(items)
ctx.execute_query()
```


The idea is to hit `items` endpoint instead of `files` endpoint to _retrieve list items with associated files and folders_, maybe this way Listview Threshold error will no longer occur.

Note: `library` refers to List resource


p.s. and if you are interested, propose to keep this issue open until the end of this month, we could verify together how to bypass Listview Threshold error for files endpoint once I'm back from vacation

**Comment by vgrem** [comment]
@mhghodsi, as promised, here is an example on how to retrieve all the files and folders within a library


```
    target_list = ctx.web.lists.get_by_title(library_name)
    items = target_list.items.select(["FileSystemObjectType"]).expand(["File", "Folder"])
    ctx.load(items)
    ctx.execute_query()
    for item in items:
        if item.properties["FileSystemObjectType"] == FileSystemObjectType.Folder:
            print("Folder url: {0}".format(item.folder.serverRelativeUrl))
        else:
            print("File url: {0}".format(item.file.serverRelativeUrl))  

```

Notes: 

- the query  `target_list.items.select(["FileSystemObjectType"]).expand(["File", "Folder"])`  construct a request to retrieve items along with associated `Folder `and `Files `properties

- `FileSystemObjectType` property allows to determine whether the item is a `Folder` or a `File`

**Comment by pecalleja** [comment]
And how can we use this but filter over a specific folder ?
